our latest

Adirl opened the pull request.

DanielBar-On committed:
msg/async/rdma: add rdma local configuration for systemd

issue: 992610

The script is only called if rdma is found in ceph.conf.

If ceph-conf or the script return an error, the comand/script is
ignored and won't affect the service, since both ExecStartPre and
ExecStopPost commands begin with the '-' sign.

Service is only reloaded if a change was made to any of the conf files.

No changes are kept after rdma is removed and services restarted (the
script restores previous configuration when stopping the service).

All changes are made in /run/systemd/system/ where all files are
considered temporary and are  only kept for the current run of the service.

Signed-off-by: DanielBar-On <danielbo@mellanox.com>
Change-Id: I57d30cc4a3a828282cd65ac2df4c8302e063c0dc

Conflicts:
	PendingReleaseNotes

Commit:
msg/async/rdma: improves RX buffer management

The commit adds following changes:
- rx buffers are allocated from the memory pool (boost::pool)
- flat memory layout for buffer data/metadata to reduce cpu cache misses
- number of receive buffers can be much larger than receive queue len
- post new buffers to the srq as soon as possible.
- stat counters

saritz committed:
msg/async/rdma: Update fix compilation warning

Assert on last errno from write function. Because lderr can ovvrride
last errno

issue: 990480

Change-Id: I2e1b08bdd8c8555816031c0ef74dbbeb1b9475d9
Signed-off-by: Sarit Zubakov <saritz@mellanox.com>

Conflicts:
	src/msg/async/rdma/RDMAConnectedSocketImpl.cc

DanielBar-On committed:
msg/async/rdma: automatically set RDMAV_HUGEPAGES_SAFE according to conf

Signed-off-by: DanielBar-On <danielbo@mellanox.com>

DanielBar-On committed:
msg/async/rdma: remove assert from ibv_dealloc_pd in ProtectionDomain
destructor

issue: 1044307

Change-Id: I2fe2099c5f4403f14877e0fbe5d4868dd34d4f9b
Signed-off-by: DanielBar-On <danielbo@mellanox.com>

Commit:
msg/async/rdma: if failed to send push csi to the back

Change-Id: If6c72917b6e3a836279b86a4d5e9e54383d7b4bd
Signed-off-by: Adir lev <adirl@mellanox.com>

Commit:
msg/async/rdma: allocated reged mem can be greater than requires len

Change-Id: I3ec42afca39c2ad93fc66ed69b3609b68f315e34
Signed-off-by: Adir lev <adirl@mellanox.com>

Commit:
msg/async/rdma: avoid adding csi twice to pending_list

Change-Id: Id76ffbfebc54c2da339e0e7f063c5420042fbe7c
Signed-off-by: Adir lev <adirl@mellanox.com>

Commit:
msg/async/rdma: avoid adding worker twice to pending_workers

Change-Id: I4dda9040a055bdb8787ffd23a6928925d0a69d83
Signed-off-by: Adir Lev <adirl@mellanox.com>

Commit:
msg/async/rdma: register buffer as continuous

Change-Id: I1a715de28ace3ff8ae077ca783ad5060e7966d54
Signed-off-by: Adir Lev <adirl@mellanox.com>

alex-mikheev committed:
msg/async/rdma: fixes hung during exit

Properly stops dispatcher polling thread. Fixes bug introduced in ec2e6931

Commit:
msg/async/rdma: create infiniband per RDMAStack

Signed-off-by: Adir lev <adirl@mellanox.com>

Adirl commented:
@alex-mikheev pls take a look

alex-mikheev commented on src/msg/async/rdma/RDMAStack.cc:
```diff
@@ -453,9 +451,9 @@ int RDMAWorker::listen(entity_addr_t &sa, const SocketOptions &opt,ServerSocket
+  stack->get_infiniband()->init();
 
-  RDMAConnectedSocketImpl* p = new RDMAConnectedSocketImpl(cct, global_infiniband.get(), get_stack()->get_dispatcher(), this);
+  RDMAConnectedSocketImpl* p = new RDMAConnectedSocketImpl(cct, stack->get_infiniband(), get_stack()->get_dispatcher(), this);
```
should be get_stack()->get_infiniband() everywhere in RDMAWorker::

Adirl replied:
because private member? 
(code compiles)

alex-mikheev replied:
it is a coding style. Either use get_stack() or stack-> and get rid of get_stack() accessor

Adirl replied:
and  RDMADispatcher::  ?

alex-mikheev commented on src/msg/async/rdma/RDMAStack.cc:
```diff
@@ -511,36 +511,37 @@ void RDMAWorker::handle_pending_message()
-  //
-  //On RDMA MUST be called before fork
-  //
+static void RDMAStack::verify_prereq(CephContext *cct) {
```
need to verify that the function is only called once.

alex-mikheev commented on src/msg/async/rdma/RDMAStack.cc:
```diff
@@ -543,12 +541,10 @@ RDMAStack::RDMAStack(CephContext *cct, const string &t): NetworkStack(cct, t)
+  ldout(cct, 20) << __func__ << " constructing Infiniband..." << dendl;
   dispatcher = new RDMADispatcher(cct, this);
-  global_infiniband->set_dispatcher(dispatcher);
+  ib->set_dispatcher(dispatcher);
```
i think there is no need to keep a dispatcher pointer in ib now. Everyone can just get it from the stack

alex-mikheev commented on src/msg/async/rdma/RDMAStack.cc:
```diff
@@ -543,12 +541,10 @@ RDMAStack::RDMAStack(CephContext *cct, const string &t): NetworkStack(cct, t)
-    global_infiniband.construct(
-      cct, cct->_conf->ms_async_rdma_device_name, cct->_conf->ms_async_rdma_port_num);
-  ldout(cct, 20) << __func__ << " constructing RDMAStack..." << dendl;
+  ib = new Infiniband(cct, cct->_conf->ms_async_rdma_device_name, cct->_conf->ms_async_rdma_port_num);
```
imho you can just pass a cephcontext to the infiniband constructor 
Also you can make it a member of rdmastack instead of using new/delete

Commit:
msg/async/rdma: remove dispatcher from infiniband

Signed-off-by: Adir lev <adirl@mellanox.com>

Commit:
msg/async/rdma: change Infiniband ctor params

Signed-off-by: Adir lev <adirl@mellanox.com>

sashakot commented on PendingReleaseNotes:
```diff
@@ -171,3 +171,7 @@
   enumerate all objects along with their snapshot and clone metadata.
+
+* When using ceph systemd services with RDMA, a script is run that adds/removes
+  configurations to the systemd service file. Therefore,the rdma option in the
```
A space between "Therefore" and "the" is absent.

sashakot commented on src/msg/async/rdma/Infiniband.cc:
```diff
@@ -470,8 +470,7 @@ Infiniband::ProtectionDomain::ProtectionDomain(CephContext *cct, Device *device)
 {
-  int rc = ibv_dealloc_pd(pd);
-  assert(rc == 0);
+  ibv_dealloc_pd(pd);
```
Why the assert was removed?

alex-mikheev replied:
Only commits starting from the https://github.com/Mellanox/ceph/pull/1/commits/793d73937fe51b521b9012ee61ebd5e18f45f0b2 are relevant

assert() is removed because it is possible to get a bad exit code (resources busy) if client app is terminated by the signal (Ctrl-C). It is easier to let OS do a cleanup than trying to do a proper cleanup in a signal handler

sashakot commented on src/msg/async/rdma/Infiniband.cc:
```diff
@@ -679,15 +733,30 @@ void Infiniband::MemoryManager::free_huge_pages(void *ptr)
   if (real_size != 0)
     munmap(real_ptr, real_size);
```
Return value of "munmap" should be checked

Commit:
msg/async/rdma check prereq only once

Signed-off-by: Adir lev <adirl@mellanox.com>

Commit:
msg/async/rdma: infiniband as member of RDMAStack

Change-Id: I0a61cc1f7aabfa3f42951838bfa7045c71982f35
Signed-off-by: Adir Lev <adirl@mellanox.com>

Commit:
fixes

Change-Id: Iccd00b255d52ef2bff2d0c275633fcd75b43bc1e
Signed-off-by: Adir Lev <adirl@mellanox.com>

Commit:
dispatcher as member of RDMAStack

Change-Id: Iccedcde3717651ced26c7402b841873d64135e53
Signed-off-by: Adir Lev <adirl@mellanox.com>

alex-mikheev committed:
msg/async/rdma: refactor rx buffer pool allocator

Pool allocator now has a context. Each context can have a
different configuration. It allows to use buffer pool safely
when there are multiple RDMAStack running in parrallel.

Do not fail on assertion when out of memory for rx buffers.
Instead log warning and try to recover.

alex-mikheev committed:
msg/async/rdma: refactor rx buffer pool allocator: CR fixes

alex-mikheev committed:
msg/async/rdma: refactor rx buffer pool allocator: CR fixes p2

out of rx buffers error message updated

Commit:
remove inflight

Signed-off-by: Adir lev <adirl@mellanox.com>

alex-mikheev reviewed (approved):
👍
